### PR TITLE
init commit

### PR DESCRIPTION
**Pull request by CreatingTheNewWorld**
init commit

### CODE REVIEW

File: android-sdk/src/main/java/ai/evolv/android_sdk/Options.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
+package ai.evolv.android_sdk;
+
+import ai.evolv.android_sdk.httpclients.HttpClient;
+
+
+
+public class Options {
```
Comment (robertsevernsentient) [inline android-sdk/src/main/java/ai/evolv/android_sdk/Options.java:7]: Not clear what this class is doing - all commented out?

**Review by robertsevernsentient — CHANGES_REQUESTED** [review]

File: android-sdk/src/main/java/ai/evolv/android_sdk/EventEmitter.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    void contaminate(JsonArray allocations)  {
+        sendAllocationEvents(CONTAMINATE_KEY, allocations);
+    }
+
+    void sendAllocationEvents(String key, JsonArray allocations) {
+        for (JsonElement a : allocations) {
+            JsonObject allocation = a.getAsJsonObject();
+            if (!audience.filter(participant.getUserAttributes(), allocation) && Allocations.isTouched(allocation)
```
[the remark this answers is not in the document]
Comment (CreatingTheNewWorld) [reply android-sdk/src/main/java/ai/evolv/android_sdk/EventEmitter.java:58]: this part was used from an old project. There is no such thing in the new one. Thanks for the explanation

File: app/src/test/java/ai/evolv/example/ExampleUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+/**
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
+ */
+public class ExampleUnitTest {
```
[the remark this answers is not in the document]
Comment (CreatingTheNewWorld) [reply app/src/test/java/ai/evolv/example/ExampleUnitTest.java:12]: this is the default class that is created automatically when creating any android project. 
I will fill it with Unit tests

File: app/src/main/java/ai/evolv/example/CustomAllocationStore.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package ai.evolv.example;
+
+import com.google.gson.JsonArray;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import ai.evolv.android_sdk.evolvinterface.EvolvAllocationStore;
+
+public class CustomAllocationStore implements EvolvAllocationStore {
```
[the remark this answers is not in the document]
Comment (CreatingTheNewWorld) [reply app/src/main/java/ai/evolv/example/CustomAllocationStore.java:10]: was removed

File: app/src/androidTest/java/ai/evolv/example/ExampleInstrumentedTest.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+import static org.junit.Assert.*;
+
+/**
+ * Instrumented test, which will execute on an Android device.
+ *
+ * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
+ */
+@RunWith(AndroidJUnit4.class)
+public class ExampleInstrumentedTest {
```
[the remark this answers is not in the document]
Comment (CreatingTheNewWorld) [reply app/src/androidTest/java/ai/evolv/example/ExampleInstrumentedTest.java:19]: this is the default class that is created automatically when creating any android project

File: android-sdk/src/test/java/ai/evolv/android_sdk/ExampleUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+/**
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
+ */
+public class ExampleUnitTest {
```
[the remark this answers is not in the document]
Comment (CreatingTheNewWorld) [reply android-sdk/src/test/java/ai/evolv/android_sdk/ExampleUnitTest.java:12]: this is the default class that is created automatically when creating any android project.
of course, I will create tests in the SDK itself, tnx

File: android-sdk/src/main/java/ai/evolv/android_sdk/EvolvClientImpl.java [hunk 7]
```diff
@@ -0,0 +1,173 @@
+package ai.evolv.android_sdk;
```
[the remark this answers is not in the document]
Comment (CreatingTheNewWorld) [reply android-sdk/src/main/java/ai/evolv/android_sdk/EvolvClientImpl.java:1]: sure, i will

File: android-sdk/src/main/java/ai/evolv/android_sdk/Execution.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        this.function = function;
+        this.participant = participant;
+        this.store = store;
+    }
+
+    String getKey() {
+        return key;
+    }
+
+    void executeWithAllocation(JsonArray rawAllocations) throws EvolvKeyError {
```
[the remark this answers is not in the document]
Comment (CreatingTheNewWorld) [reply android-sdk/src/main/java/ai/evolv/android_sdk/Execution.java:36]: was removed

File: android-sdk/src/main/java/ai/evolv/android_sdk/ExecutionQueue.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
+package ai.evolv.android_sdk;
```
[the remark this answers is not in the document]
Comment (CreatingTheNewWorld) [reply android-sdk/src/main/java/ai/evolv/android_sdk/ExecutionQueue.java:1]: sure, i will

File: android-sdk/src/main/java/ai/evolv/android_sdk/Options.java (hunk 1 above)
Comment (CreatingTheNewWorld) [reply android-sdk/src/main/java/ai/evolv/android_sdk/Options.java:7]: was removed

File: android-sdk/src/main/java/ai/evolv/android_sdk/EvolvContextImpl.java [hunk 10]
```diff
@@ -36,17 +52,75 @@ private void initialize(String uid,
...
+        }
+
+        ensureInitialized();
+
+        JsonElement context = local ? localContext : remoteContext;
+        // TODO: 01.06.2021 add "before"
 
+        helper.setKeyToValue(key, jsonValue, context);
+
+        Object updated = this.resolve();
```
Comment (robertsevernsentient) [inline android-sdk/src/main/java/ai/evolv/android_sdk/EvolvContextImpl.java:100]: This is unused?

File: android-sdk/src/main/java/ai/evolv/android_sdk/EvolvContextImpl.java [hunk 11]
```diff
@@ -36,17 +52,75 @@ private void initialize(String uid,
...
+            //this.localContext = new HashMap<>();
+        }
+
         initialized = true;
 
-        // TODO: 25.05.2021 emit
-        //emit(this, CONTEXT_INITIALIZED, this.resolve());
+        Object resolve = resolve();
+        List<Object> objects = new ArrayList<>();
+        objects.add(CONTEXT_INITIALIZED);
```
Comment (robertsevernsentient) [inline android-sdk/src/main/java/ai/evolv/android_sdk/EvolvContextImpl.java:74]: Does this need to be added to the object given it's already the 2nd argument in the emit?

File: android-sdk/src/main/java/ai/evolv/android_sdk/EvolvContextImpl.java [hunk 12]
```diff
@@ -36,17 +52,75 @@ private void initialize(String uid,
...
+        // TODO: 01.06.2021 understand the data type  emit(,,?)
+        waitForIt.emit(this, CONTEXT_CHANGED, null);
+        return true;
+
+    }
+
+    @Override
+    public Object resolve() {
+        ensureInitialized();
+        //return objects.deepClone(mutableResolve());
```
Comment (robertsevernsentient) [inline android-sdk/src/main/java/ai/evolv/android_sdk/EvolvContextImpl.java:113]: What's this bit doing?

File: android-sdk/src/main/java/ai/evolv/android_sdk/EvolvEmitter.java [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    public EvolvEmitter(String endpoint, EvolvContext evolvContext, boolean blockTransmit) {
+
+        this.endpoint = endpoint;
+        this.evolvContext = evolvContext;
+        this.blockTransmit = blockTransmit;
+
+    }
+
+    void emit(String type, Map<String, List<Object>> payload, boolean flush){
+
```
Comment (robertsevernsentient) [inline android-sdk/src/main/java/ai/evolv/android_sdk/EvolvEmitter.java:25]: Please keep track of the unimplemented methods with a common TODO note

File: android-sdk/src/main/java/ai/evolv/android_sdk/EvolvStoreImpl.java [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+
+        Iterator<JsonElement> iterator = allocs.iterator();
+
+        while (iterator.hasNext()) {
+            JsonObject allocObject = iterator.next().getAsJsonObject();
+
+            if (allocObject.has(GENOME_STRING)) {
+                allocObject.remove(GENOME_STRING);
+            }
+            if (allocObject.has(AUDIENCE_QUERY_STRING)) {
```
Comment (robertsevernsentient) [inline android-sdk/src/main/java/ai/evolv/android_sdk/EvolvStoreImpl.java:278]: What's 'audience_query'?

File: android-sdk/src/main/java/ai/evolv/android_sdk/Execution.java [hunk 15]
```diff
@@ -34,21 +34,7 @@ String getKey() {
     }
 
     void executeWithAllocation(JsonArray rawAllocations) throws EvolvKeyError {
```
Comment (robertsevernsentient) [inline android-sdk/src/main/java/ai/evolv/android_sdk/Execution.java:36]: Is this an empty function now? If so please add TODO to clean up

File: android-sdk/src/main/java/ai/evolv/android_sdk/helper/UtilityHelper.java [hunk 16]
```diff
@@ -13,4 +21,26 @@
             return null;
         }
     }
+
+    public JsonElement setKeyToValue(String key, JsonElement value, JsonElement map){
```
Comment (robertsevernsentient) [inline android-sdk/src/main/java/ai/evolv/android_sdk/helper/UtilityHelper.java:25]: Me ntioned unit test suite before, but something like this really needs tests

File: android-sdk/src/main/java/ai/evolv/android_sdk/helper/UtilityHelper.java [hunk 17]
```diff
@@ -13,4 +21,26 @@
...
+
+        JsonObject current = (JsonObject) map;
+        String[] keys = key.split(Pattern.quote("."));
+        for (int i = 0; i < keys.length; i++) {
+            String k = keys[i];
+            if (i == (keys.length - 1)) {
+                current.add(k,value);
```
Comment (robertsevernsentient) [inline android-sdk/src/main/java/ai/evolv/android_sdk/helper/UtilityHelper.java:32]: Does 'add' work is the key is already present?

File: android-sdk/src/main/java/ai/evolv/android_sdk/WaitForIt.java [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+import java.util.Set;
+
+import ai.evolv.android_sdk.evolvinterface.EvolvAction;
+import ai.evolv.android_sdk.evolvinterface.EvolvContext;
+import ai.evolv.android_sdk.evolvinterface.EvolvInvocation;
+
+import static ai.evolv.android_sdk.EvolvContextImpl.CONTEXT_VALUE_ADDED;
+import static ai.evolv.android_sdk.EvolvContextImpl.CONTEXT_VALUE_REMOVED;
+
+class WaitForIt {
```
Comment (robertsevernsentient) [inline android-sdk/src/main/java/ai/evolv/android_sdk/WaitForIt.java:20]: Please add the TODO to get this implemented -- also, very testable this class

File: android-sdk/src/main/java/ai/evolv/android_sdk/EvolvContextImpl.java (hunk 10 above)
Comment (CreatingTheNewWorld) [reply android-sdk/src/main/java/ai/evolv/android_sdk/EvolvContextImpl.java:100]: photo_2021-06-03_11-12-29

File: android-sdk/src/main/java/ai/evolv/android_sdk/EvolvContextImpl.java (hunk 11 above)
Comment (CreatingTheNewWorld) [reply android-sdk/src/main/java/ai/evolv/android_sdk/EvolvContextImpl.java:74]: in Java, the method signature is strictly defined.
that is, when we need to pass a different number of parameters to one method, we either make a list or use the @overload annotation
photo_2021-06-03_11-29-27_1
photo_2021-06-03_11-29-19_2

File: android-sdk/src/main/java/ai/evolv/android_sdk/EvolvContextImpl.java (hunk 12 above)
Comment (CreatingTheNewWorld) [reply android-sdk/src/main/java/ai/evolv/android_sdk/EvolvContextImpl.java:113]: it is the same logic as "**objects.deepClone(mutableResolve()**);"
photo_2021-06-03_11-32-37
photo_2021-06-03_11-33-47

File: android-sdk/src/main/java/ai/evolv/android_sdk/EvolvEmitter.java (hunk 13 above)
Comment (CreatingTheNewWorld) [reply android-sdk/src/main/java/ai/evolv/android_sdk/EvolvEmitter.java:25]: sure, tnx u

File: android-sdk/src/main/java/ai/evolv/android_sdk/EvolvStoreImpl.java (hunk 14 above)
Comment (CreatingTheNewWorld) [reply android-sdk/src/main/java/ai/evolv/android_sdk/EvolvStoreImpl.java:278]: 'audience_query' - string variable 
photo_2021-06-03_11-41-145
photo_2021-06-03_11-44-505
photo_2021-06-03_11-45-375

File: android-sdk/src/main/java/ai/evolv/android_sdk/Execution.java (hunk 15 above)
Comment (CreatingTheNewWorld) [reply android-sdk/src/main/java/ai/evolv/android_sdk/Execution.java:36]: done
tnx

File: android-sdk/src/main/java/ai/evolv/android_sdk/helper/UtilityHelper.java (hunk 16 above)
Comment (CreatingTheNewWorld) [reply android-sdk/src/main/java/ai/evolv/android_sdk/helper/UtilityHelper.java:25]: I will make unit tests for it 
tnx

File: android-sdk/src/main/java/ai/evolv/android_sdk/helper/UtilityHelper.java (hunk 17 above)
Comment (CreatingTheNewWorld) [reply android-sdk/src/main/java/ai/evolv/android_sdk/helper/UtilityHelper.java:32]: photo_2021-06-03_11-57-28

File: android-sdk/src/main/java/ai/evolv/android_sdk/WaitForIt.java (hunk 18 above)
Comment (CreatingTheNewWorld) [reply android-sdk/src/main/java/ai/evolv/android_sdk/WaitForIt.java:20]: sure

**Review by robertsevernsentient — APPROVED** [review]

**Pull request merged**